src/relax/transform/canonicalize_bindings.cc

```diff
@@ -102,6 +103,108 @@ class BindingCanonicalizer : public ExprMutator {
     }
   }
 
+  // Special case: for dataflow blocks, we will check for dataflow vars that solely exist
+  // to be bound to the output. In this case, we will get rid of those bindings and
+  // use the dataflow var's definition directly
+  BindingBlock VisitBindingBlock_(const DataflowBlockNode* block) override {
+    auto new_block = Downcast<DataflowBlock>(ExprMutator::VisitBindingBlock_(block));
+    std::unordered_map<DataflowVar, Expr, ObjectPtrHash, ObjectPtrEqual> binding_map;
+    std::unordered_set<DataflowVar, ObjectPtrHash, ObjectPtrEqual> disqualified_set;
+    std::unordered_set<DataflowVar, ObjectPtrHash, ObjectPtrEqual> output_vars;
+
+    for (auto binding : new_block->bindings) {
```

Comment:
Whether a variable is a potential "output" variable, and whether it's disqualified from that depends on how it's used.  It's easier to collect that information going backwards in the dataflow order.

Reply:
Good point